build: TypeScript 3.6 compatibility.

mprobst opened the pull request:
This PR updates Angular to compile with TypeScript 3.6 while retaining
compatibility with TS3.5. We achieve this by inserting several `as any`
casts for compatiblity around `ts.CompilerHost` APIs.

mprobst committed:
build: TypeScript 3.6 compatibility.

This PR updates Angular to compile with TypeScript 3.6 while retaining
compatibility with TS3.5. We achieve this by inserting several `as any`
casts for compatiblity around `ts.CompilerHost` APIs.

mprobst commented:
@kara the area really is `all`, but that doesn't appear to be a category.

LMK if this works for you with the `any`s. I believe it'd also be possible to introduce some indirection of the TS types for the backwards compat, but that'd require cross-component dependencies and IMHO wouldn't really buy us much, compared to the few call sites.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/commonjs_host.ts:
```diff
@@ -180,8 +180,9 @@ export class CommonJsReflectionHost extends Esm5ReflectionHost {
   private resolveModuleName(moduleName: string, containingFile: ts.SourceFile): ts.SourceFile
       |undefined {
     if (this.compilerHost.resolveModuleNames) {
+      // FIMXE: remove the "as any" cast once on TS3.6.
```
```suggestion
      // FIXME: remove the "as any" cast once on TS3.6.
```

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/umd_host.ts:
```diff
@@ -158,8 +158,9 @@ export class UmdReflectionHost extends Esm5ReflectionHost {
   private resolveModuleName(moduleName: string, containingFile: ts.SourceFile): ts.SourceFile
       |undefined {
     if (this.compilerHost.resolveModuleNames) {
+      // FIMXE: remove the "as any" cast once on TS3.6.
```
```suggestion
      // FIXME: remove the "as any" cast once on TS3.6.
```

petebacondarwin commented on packages/compiler-cli/src/metadata/collector.ts:
```diff
@@ -458,7 +458,7 @@ export class MetadataCollector {
                     expression: recordEntry({__symbolic: 'reference', name: enumName}, node), name
                   } as any,
-                };
+                } as any;
```
No FIXME comment here?

mprobst replied:
I think it's the same as before here, really, we're just moving the `any` up a level. So I think that's just how the code operates? TS 3.6 appears to bail on an object literal with an any member here, whereas 3.5 allowed it.

petebacondarwin replied:
So perhaps remove the inner `as any`?

mprobst replied:
Done.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/util/src/typescript.ts:
```diff
@@ -113,7 +113,9 @@ export function resolveModuleName(
   if (compilerHost.resolveModuleNames) {
-    return compilerHost.resolveModuleNames([moduleName], containingFile)[0];
+    // FIXME: Additional parameters are required in TS3.6, but ignored in 3.5.
+    // Remove the any case once fully on TS3.6.
```
```suggestion
    // Remove the any cast once fully on TS3.6.
```

petebacondarwin commented on packages/core/src/application_ref.ts:
```diff
@@ -277,7 +277,7 @@ export class PlatformRef {
           {providers: providers, parent: this.injector, name: moduleFactory.moduleType.name});
       const moduleRef = <InternalNgModuleRef<M>>moduleFactory.create(ngZoneInjector);
-      const exceptionHandler: ErrorHandler = moduleRef.injector.get(ErrorHandler, null);
+      const exceptionHandler: ErrorHandler|null = moduleRef.injector.get(ErrorHandler, null);
```
How come our strict null checks did not catch this before? Or is `injector.get()` typed not to return a `null` value?

mprobst replied:
I think it resolved the wrong overload for calling `.get(T, null)`. We define:

```
get<T>(
      token: Type<T>|InjectionToken<T>|AbstractType<T>, notFoundValue?: T, flags?: InjectFlags): T;
```

So as we pass `ErrorHandler` for `token`, and `null` for `notFoundValue`, 3.5 used the first value, whereas 3.6 merges the two values into a union of `ErrorHandler|null`.

We could also change the definition (but that'd be an API change):

```
get<T, U>(
      token: Type<T>|InjectionToken<T>|AbstractType<T>, notFoundValue?: U, flags?: InjectFlags): T|U;
```

And then if `T === U`, TS should collapse the union to just `T`.

petebacondarwin replied:
Fair enough - it doesn't warrant a public API change.

petebacondarwin commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -81,7 +81,7 @@ export interface DirectiveType<T> extends Type<T> {
 }
 
-export const enum DirectiveDefFlags {ContentQuery = 0b10}
+export enum DirectiveDefFlags {ContentQuery = 0b10}
```
Why remove the `const`?

mprobst replied:
Const enums are guaranteed to be fully removed and inlined to their members. But we re-export the enum, which means the object would need to be retained. Here's the error:

```
 error TS2475: 'const' enums can only be used in property or index access expressions or the right hand side of an import declaration or export assignment or type query.
```

gkalpak replied:
I wonder if this is used at all. I couldn't find any reference to `DirectiveDefFlags` (other than declaring it and exporting it) and it doesn't seem to be part of the [public API either][1]

[1]: https://next.angular.io/?search=DirectiveDefFlags

IgorMinar replied:
I dropped this in #32946

gkalpak reviewed (approved):
LGTM (except for the commit messages, which need to follow the [guidelines][1] to make the linter happy) ❤️ 

[1]: https://github.com/angular/angular/blob/877f9ad7b9151324c4a94d3c80e7d37691bb9fd5/CONTRIBUTING.md#commit

mprobst commented:
@gkalpak do you mean the commit msg from the subsequent fixups, or the original one? I hope the original one is OK, let me know if it's not.

I've rebased onto master and squashed.

gkalpak commented:
I meant the subsequent commits (which are gone now). But now that you mentioned it, the original commit should probably not mention `compiler-cli` as the scope, since changes are made to other packages as well :grin:

Also, the formatter complains about misformatted files now. (You can see the error [here][1].)
You can fix it with `gulp format`.

##
BTW, it is outside the scope of this PR, but mentioning it "for completeness":
There is some more work needed to make Angular officially support TS3.6. (See, for example, the equivalent PR for adding support for TS3.5: #31615)

[1]: https://circleci.com/gh/angular/angular/467093

mprobst commented:
@gkalpak done.

IgorMinar reviewed (approved):
lgtm, I created a follow up PR that properly upgrades to tsc 3.6 and cleans up some of Martin's changes: #32946 this PR is blocked on typscript 3.6.4 release, so I think we should go ahead and merge this one in to unblock tsc update in google3, while we wait on the typescript release.

mprobst commented:
@IgorMinar I've rebased this onto latest master.

IgorMinar commented:
merge-assistance: global approval

I also kicked off a presubmit: http://test/OCL:272419562:BASE:272419580:1570020811421:232c8db4
